### PR TITLE
Lightning integration

### PR DESCRIPTION
**Pull request by szmazurek**
Fixes #

## Proposed Changes
- added lightinign-based logic to the training and inference managers
- removed bugs found on the way
- conducted testing and debugging - still, some edge cases remain that are causing some tests to fail, altoguht they semm quicly fixable

## Checklist

- [ ] [`CONTRIBUTING`](https://github.com/mlcommons/GaNDLF/blob/master/CONTRIBUTING.md) guide has been followed.
- [ ] PR is based on the [current GaNDLF master ](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch-in-github-desktop?platform=windows).
- [ ] Non-breaking change (does **not** break existing functionality): provide **as many** details as possible for _any_ breaking change.
- [ ] Function/class source code documentation added/updated (ensure `typing` is used to provide type hints, including and not limited to using `Optional` if a variable has a pre-defined value).
- [ ] Code has been [blacked](https://github.com/psf/black#usage) for style consistency and linting.
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/version.py).
- [ ] If adding a git submodule, add to list of exceptions for black styling in [pyproject.toml](https://github.com/mlcommons/GaNDLF/blob/master/pyproject.toml) file.
- [ ] [Usage documentation](https://github.com/mlcommons/GaNDLF/blob/master/docs) has been updated, if appropriate.
- [ ] Tests added or modified to [cover the changes](https://app.codecov.io/gh/mlcommons/GaNDLF); if coverage is reduced, please give explanation.
- [ ] If customized dependency installation is required (i.e., a separate `pip install` step is needed for PR to be functional), please ensure it is reflected in all the files that control the CI, namely: [python-test.yml](https://github.com/mlcommons/GaNDLF/blob/master/.github/workflows/python-test.yml), and all docker files [[1](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-CPU),[2](https://github.com/mlcommons/GaNDLF/blob/devcontainer_build_fix/Dockerfile-CUDA11.6),[3](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-ROCm)].
- [ ] The `logging` library is being used and no `print` statements are left.

### CODE REVIEW

File: GANDLF/metrics/metric_calculators.py [hunk 1]
```diff
@@ -13,7 +14,9 @@ def __init__(self, params: dict):
     def _initialize_metrics_dict(self):
         self.metrics_calculators = get_metrics(self.params)
 
-    def _process_metric_value(self, metric_value: torch.Tensor):
+    def _process_metric_value(self, metric_value: Union[torch.Tensor, float]):
+        if isinstance(metric_value, float):
+            return metric_value
```
Comment (sarthakpati) [inline GANDLF/metrics/metric_calculators.py:19]: Should the bottom `if` be `ifelse`?

File: samples/config_all_options.yaml [hunk 2]
```diff
@@ -95,7 +95,7 @@ num_epochs: 100
 patience: 50
 # Set the batch size
 batch_size: 1
-# gradient clip : norm, value, agc
+# gradient clip : norm, value
```
Comment (sarthakpati) [inline samples/config_all_options.yaml:98]: Since [Lightning has its own gradient clipping](https://lightning.ai/docs/pytorch/stable/common/optimization.html#gradient-clipping), Do we need our custom [gradient clipping code](https://github.com/mlcommons/GaNDLF/tree/master/GANDLF/grad_clipping) anymore?

File: samples/config_all_options.yaml [hunk 3]
```diff
@@ -304,3 +304,7 @@ q_samples_per_volume: 5
 q_num_workers: 2 # scale this according to available CPU resources
 # used for debugging
 q_verbose: False
+
+accelerator: "cpu" # options: "cpu", "gpu", "auto", "auto" selects GPU if available
```
Comment (sarthakpati) [inline samples/config_all_options.yaml:308]: These should be part of docs.

File: testing/test_full.py [hunk 4]
```diff
@@ -2913,6 +2910,7 @@ def test_generic_deploy_docker():
     print("passed")
 
 
+@pytest.mark.skip("Differential privacy not implemented")
```
Comment (sarthakpati) [inline testing/test_full.py]: ??? Does Lightning not support Opacus?

File: samples/config_all_options.yaml (hunk 2 above)
Comment (szmazurek) [reply samples/config_all_options.yaml:98]: Not really, should I remove it too?

File: samples/config_all_options.yaml (hunk 3 above)
Comment (szmazurek) [reply samples/config_all_options.yaml:308]: Right, will add

File: testing/test_full.py (hunk 4 above)
Comment (szmazurek) [reply testing/test_full.py]: It does, but I purposefully skipped during implementation now, as I wanted to have base implementation working now. Do we also need it for the current port?

Comment (sarthakpati) [reply testing/test_full.py]: It would be good to have that included as well, since there are active experiments that utilize this. Could be another benchmarking test on our (primarily my) end as well.

Comment (szmazurek) [reply testing/test_full.py]: added

File: GANDLF/metrics/metric_calculators.py (hunk 1 above)
Comment (szmazurek) [reply GANDLF/metrics/metric_calculators.py:19]: Could be but, not necessairly, each statement will be checked anyway - this method will return something, if we missed any case then it means it is a bug.

File: samples/config_all_options.yaml (hunk 3 above)
Comment (szmazurek) [reply samples/config_all_options.yaml:308]: Added

File: samples/config_all_options.yaml (hunk 2 above)
Comment (szmazurek) [reply samples/config_all_options.yaml:98]: Removed

File: GANDLF/_graveyard/README.md
Comment (sarthakpati) [inline GANDLF/_graveyard/README.md:1]: I would recommend renaming the `_graveyard` submodule to `_deprecated`

Comment (szmazurek) [reply GANDLF/_graveyard/README.md:1]: ok, good idea

File: GANDLF/compute/generic.py [hunk 5]
```diff
@@ -7,42 +7,49 @@
...
-        create_subset_dataloader(), as this method will populate the parameters
+        create_subset_dataset(), as this method will populate the parameters
         dict with the headers from the subset data.
         """
         return self.parameters_dict
 
 
 class TrainingSubsetDataParser(AbstractSubsetDataParser):
-    def create_subset_dataloader(self) -> DataLoader:
+    def create_subset_dataset(self):
```
Comment (sarthakpati) [inline GANDLF/compute/generic.py]: Return typing hint could be useful.

File: GANDLF/compute/generic.py [hunk 6]
```diff
@@ -64,31 +71,44 @@ def create_subset_dataloader(self) -> DataLoader:
...
+            get_dataframe(self.parameters_dict["training_data"]),
+            self.parameters_dict,
+            train=True,
+            loader_type="train",
+        )
 
 
 class ValidationSubsetDataParser(AbstractSubsetDataParser):
-    def create_subset_dataloader(self) -> DataLoader:
+    def create_subset_dataset(self):
```
Comment (sarthakpati) [inline GANDLF/compute/generic.py]: Return typing hint could be useful.

File: GANDLF/compute/generic.py [hunk 7]
```diff
@@ -64,31 +71,44 @@ def create_subset_dataloader(self) -> DataLoader:
...
+        )
+        return validation_dataset
 
 
 class TestSubsetDataParser(AbstractSubsetDataParser):
-    def create_subset_dataloader(self) -> DataLoader:
-        # hack to cope with the get_testing_loader() function
-        self.parameters_dict["testing_data"] = self.subset_csv_path
-        return get_testing_loader(self.parameters_dict)
+    def create_subset_dataset(self):
```
Comment (sarthakpati) [inline GANDLF/compute/generic.py]: Return typing hint could be useful.

File: GANDLF/inference_manager.py [hunk 8]
```diff
@@ -112,17 +111,15 @@ def InferenceManager(
...
 
             if parameters.get("auto_batch_size_find", False):
                 if parameters["modality"] in ["path", "histo"]:
                     print(
-                        "Auto batch size find is not supported for histo images. Using default batch size."
-                    )
-                else:
-                    LightningTuner(trainer).scale_batch_size(
-                        lightning_module, datamodule=datamodule, method="predict"
+                        "Auto batch size find is not supported in inference. Dataloader batch size is always 1."
```
Comment (sarthakpati) [inline GANDLF/inference_manager.py]: If `auto_batch_size_find` is not supporter, why is the check only for `["path", "histo"]`? Should it not be for `rad` as well?

Comment (szmazurek) [reply GANDLF/inference_manager.py]: That's correct! Missed this one

File: GANDLF/compute/generic.py (hunk 5 above)
Comment (szmazurek) [reply GANDLF/compute/generic.py]: In this case, the interpreter is able to define it automatically (as a return type from ImagesFromDataframe), so it is appearing in type hints anyway and mypy is detecting it. Do we want to write it explicitly if that is the case?

Comment (sarthakpati) [reply GANDLF/compute/generic.py]: Ah okay. I think it could be useful (from a coding perspective).

Comment (szmazurek) [reply GANDLF/compute/generic.py]: cool, so I will scan the module and it it whenever missing
